python/ray/_private/usage/usage_lib.py

```diff
@@ -0,0 +1,126 @@
+"""This is the module that is in charge of Ray usage report (telemetry) APIs.
+
+NOTE: Ray's usage report is currently "off by default".
+      But we are planning to make it opt-in by default.
+
+Ray usage report follows the specification from
+https://docs.google.com/document/d/1ZT-l9YbGHh-iWRUC91jS-ssQ5Qe2UQ43Lsoc1edCalc/edit#heading=h.17dss3b9evbj. # noqa
+
+# Module
+
+The module consists of 2 parts.
+
+## Public API
+It contains public APIs to obtain usage report information. 
+APIs will be added before the usage report becomes opt-in by default.
+
+## Internal APIs for usage processing/report
+The telemetry report consists of 5 components. This module is in charge of the top 2 layers.
+
+Report                -> usage_lib
+---------------------
+Usage data processing -> usage_lib
+---------------------
+Data storage          -> Ray API server
+---------------------
+Aggregation           -> Ray API server (currently a dashboard server)
+---------------------
+Usage data collection -> Various components (Ray agent, GCS, etc.) + usage_lib (cluster metadata).
+
+Usage report is currently "off by default". You can enable the report by setting an environment variable
+RAY_USAGE_STATS_ENABLE=1. For example, `RAY_USAGE_STATS_ENABLE=1 ray start --head`.
+Or `RAY_USAGE_STATS_ENABLE=1 python [drivers with ray.init()]`.
+
+"Ray API server (currently a dashboard server)" reports the usage data to the hard-coded URL.
+`ray._private.usage.usage_constants.USAGE_REPORT_URL`.
+
+Data is reported
+- Every `ray._private.usage.usage_constants.USAGE_REPORT_INTERVAL_S`.
+
+Note that it is also possible to configure the interval using the environment variable,
+`RAY_USAGE_REPORT_INTERVAL_S`.
+
+To see collected/reported data, see `usage_stats.json` inside a temp
+folder (e.g., /tmp/ray/session_[id]/*).
+"""
+import os
+import uuid
+import sys
+import json
+import logging
+
+
+import ray
+
+import ray.ray_constants as ray_constants
+import ray._private.usage.usage_constants as usage_constant
+
+logger = logging.getLogger(__name__)
+
+#################
+# Internal APIs #
+#################
+
+
+def _generate_cluster_metadata():
+    """Return a dictionary of cluster metadata."""
+    ray_version, python_version = ray._private.utils.compute_version_info()
+    return {
+        "schema_version": usage_constant.SCHEMA_VERSION,
+        "source": os.getenv("RAY_USAGE_STATS_SOURCE", "OSS"),
```

Comment:
confirm with @richardliaw this is the source value we want by default?

Reply:
I will merge the PR first. @richardliaw lmk what you think!